### PR TITLE
Add Pixel shuffle OP

### PR DESCRIPTION
**Pull request by shippingwang**
Add pixel shuffle Op.

Pixel shuffle rearranges elements in a tensor of shape [N, C, H, W] to [N, C / r^2, H * r, W * r], which r is an upscale_factor.

NOTE: Pixel shuffle Op. shuffles input data. NOT simple reshaping.

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -10339,3 +10340,54 @@ def tree_conv(nodes_vector,
...
+    """
+    ** Pixel Shuffle Layer**
+
+    This layer rearrange elements in a tensor of shape [N, C, H, W]
+    to a tensor of shape [N, C/r**2, H*r, W*r]
+    This is useful for implementing efficient sub-pixel convolution
+    with a stride of 1/r.
+
+    please refer to the paper: `Real-Time Single Image and Video Super-Resolution 
+    Using an Efficient Sub-Pixel Convolutional Neural Network`
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py]: give a link here?

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -10339,3 +10340,54 @@ def tree_conv(nodes_vector,
...
+            upscale_factor= 3
+        output shape is:
+            [1, 1, 12, 12]
+    
+    Args:
+        input(Variable): The input tensor variable.
+        upscale_factor(int): factor to increase spatial resolution by
+
+    Returns:
+        Out(Variable): the pixel shuffle result is a tensor variable with the same shape and the same type as the input.
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py:10956]: the -> The

**Review by heavengate — COMMENTED** [review]
Is python API test in test_layers.py needed?

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py]: OK

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py:10956]: OK

**Comment by shippingwang** [comment]
> Is python API test in test_layers.py needed?

yeah

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -10339,3 +10340,55 @@ def tree_conv(nodes_vector,
...
+    """
+    n, c, h, w = input.shape
+    if c % (upscale_factor**2) != 0:
+        raise ValueError(
+            "the square of upscale_factor must divide the channels of input")
+    return reshape(
+        x=input,
+        shape=[
+            n, c // (upscale_factor**2), h * upscale_factor, w * upscale_factor
+        ])
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: There is a problem in this API for the variable length input image.  H and W is not fixed.

**Comment by shanyi15** [comment]
plz preview it first, thanks

**Comment by shippingwang** [comment]
> plz review it first, thanks

OK will add it.

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -10923,6 +10924,65 @@ def npair_loss(anchor, positive, labels, l2_reg=0.002):
...
+def pixel_shuffle(x, upscale_factor):
+    """
+
+    ** Pixel Shuffle Layer**
+
+    This layer rearrange elements in a tensor of shape [N, C, H, W]
+    to a tensor of shape [N, C/r**2, H*r, W*r]
+    This is useful for implementing efficient sub-pixel convolution
+    with a stride of 1/r.
+    please refer to the paper: `Real-Time Single Image and Video Super-Resolution
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: please -> Please

**Comment by shippingwang** [comment]
预览不更新，尝试了很多次，不知道为什么 @shanyi15

**Comment by qingqing01** [comment]
图片

**Comment by shippingwang** [comment]
image
预览效果updated

**Review by qingqing01 — DISMISSED** [review]
LGTM.

**Comment by shanyi15** [comment]
image

**Comment by shippingwang** [comment]
image
Done!

**Comment by shippingwang** [comment]
image
DONE!

**Review by shanyi15 — DISMISSED** [review]
LG comments

**Review by qingqing01 — DISMISSED** [review]

**Review by shanyi15 — DISMISSED** [review]

**Review by panyx0718 — APPROVED** [review]

**Review by shanyi15 — APPROVED** [review]

**Pull request merged**